homeassistant/components/flux_led/light.py

```diff
@@ -139,79 +166,119 @@
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_entry(hass, entry, async_add_entities):
     """Set up the Flux lights."""
-    lights = []
-    light_ips = []
-
-    for ipaddr, device_config in config.get(CONF_DEVICES, {}).items():
-        device = {}
-        device["name"] = device_config[CONF_NAME]
-        device["ipaddr"] = ipaddr
-        device[CONF_PROTOCOL] = device_config.get(CONF_PROTOCOL)
-        device[ATTR_MODE] = device_config[ATTR_MODE]
-        device[CONF_CUSTOM_EFFECT] = device_config.get(CONF_CUSTOM_EFFECT)
-        light = FluxLight(device)
-        lights.append(light)
-        light_ips.append(ipaddr)
-
-    if not config.get(CONF_AUTOMATIC_ADD, False):
-        add_entities(lights, True)
-        return
-
-    # Find the bulbs on the LAN
-    scanner = BulbScanner()
-    scanner.scan(timeout=10)
-    for device in scanner.getBulbInfo():
-        ipaddr = device["ipaddr"]
-        if ipaddr in light_ips:
-            continue
-        device["name"] = f"{device['id']} {ipaddr}"
-        device[ATTR_MODE] = None
-        device[CONF_PROTOCOL] = None
-        device[CONF_CUSTOM_EFFECT] = None
-        light = FluxLight(device)
-        lights.append(light)
-
-    add_entities(lights, True)
+
+    async def async_new_lights(bulbs: dict):
+        """Add new bulbs when they are found or configured."""
+
+        lights = []
+
+        for bulb_id, bulb_details in bulbs.items():
+            effect_speed = entry.options.get(bulb_id, {}).get(
+                CONF_EFFECT_SPEED,
+                entry.options.get("global", {}).get(
+                    CONF_EFFECT_SPEED, DEFAULT_EFFECT_SPEED
+                ),
+            )
+
+            host = bulb_details[CONF_HOST]
+            try:
+                bulb = await hass.async_add_executor_job(WifiLedBulb, host)
+            except BrokenPipeError as error:
+                raise PlatformNotReady(error) from error
+
+            lights.append(
+                FluxLight(
+                    unique_id=bulb_id,
+                    device=bulb_details,
+                    effect_speed=effect_speed,
+                    bulb=bulb,
+                )
+            )
+
+        async_add_entities(lights, True)
+
+    await async_new_lights(entry.data[CONF_DEVICES])
+
+    async_dispatcher_connect(hass, SIGNAL_ADD_DEVICE, async_new_lights)
```

Comment:
I think we need to use the `on_ unload` helper to make sure the subscription is cancelled if the config entry is un- or reloaded, for example after the user changes options. If not, `async_new_lights` will be called multiple times and meaning we will have multiple entities for the same light.

A test to verify this is currently broken (you get multiple copies of a light) but is fixed by unsubscribing would be good.